internal/controller/controllers/agent_controller.go

```diff
@@ -72,6 +80,32 @@ func (r *AgentReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl
 		r.Log.Debugf("ClusterDeploymentName not set in Agent %s. Skipping Reconcile", agent.Name)
 		return ctrl.Result{Requeue: false}, nil
 	}
+
+	if agent.ObjectMeta.DeletionTimestamp.IsZero() { // agent not being deleted
+		// Register a finalizer if it is absent.
+		if !funk.ContainsString(agent.GetFinalizers(), AgentFinalizerName) {
+			controllerutil.AddFinalizer(agent, AgentFinalizerName)
+			if err = r.Update(ctx, agent); err != nil {
+				return ctrl.Result{}, err
```

Comment:
Maybe instead of returning an error you should log it and set requeue to true? 
Same comment to all the places that return an error.

Reply:
done